Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -270,12 +283,19 @@ void Writer::WriteTag(Tag tag) {
 
 Tag Reader::ReadTag() {
   Tag tag;
+  if (!status_.ok()) return tag;
+
   bool eof;
   bool ok = pb_decode_tag(&stream_, &tag.wire_type, &tag.field_number, &eof);
-  if (!ok || eof) {
-    // TODO(rsgowman): figure out error handling
-    abort();
+  if (!ok) {
+    status_ =
+        Status(FirestoreErrorCode::InvalidArgument, PB_GET_ERROR(&stream_));
```

Comment:
I think DataLoss is preferable because InvalidArgument implies that the client did something wrong.

Both of these are treated as perm failures though: https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Remote/FSTDatastore.mm#L165